File: frontend/services/feature-selection.ts
import { ParsedDataset } from "@/types/import";
import {
  FeatureSelectionApiResponse,
  FeatureSelectionPreview,
} from "@/types/feature-selection";

function getBackendUrl(): string {
  const backendUrl = process.env.NEXT_PUBLIC_BACKEND_URL?.replace(/\/$/, "");

  if (!backendUrl) {
    throw new Error("NEXT_PUBLIC_BACKEND_URL is not set.");
  }

  return backendUrl;
}

export async function fetchFeatureSelectionPreview(
  dataset: ParsedDataset,
  featureIndices: number[],
  targetIndex: number | null,
): Promise<FeatureSelectionPreview> {
  const response = await fetch(
    `${getBackendUrl()}/api/feature-selection/preview`,
    {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify({
        dataset: {
          format: dataset.format,
          headers: dataset.headers,
          rows: dataset.rows,
        },
        feature_indices: featureIndices,
        target_index: targetIndex,
      }),
    },
  );

  if (!response.ok) {
    throw new Error("Failed to fetch feature selection preview from backend");
  }

  const payload = (await response.json()) as FeatureSelectionApiResponse;

  return {
    format: payload.format,
    selectedHeaders: payload.selected_headers,
    featureHeaders: payload.feature_headers,
    targetHeader: payload.target_header,
    rows: payload.rows,
    summary: {
      totalColumns: payload.summary.total_columns,
      selectedFeatureCount: payload.summary.selected_feature_count,
      rowCount: payload.summary.row_count,
      previewRowCount: payload.summary.preview_row_count,
    },
  };
}

File: backend/app/api/feature_selection.py
from fastapi import APIRouter, HTTPException

from app.schemas.feature_selection import (
    FeatureSelectionRequest,
    FeatureSelectionResponse,
)
from app.services.feature_selection_service import select_features

router = APIRouter()


@router.post("/feature-selection/preview", response_model=FeatureSelectionResponse)
def feature_selection_preview(payload: FeatureSelectionRequest):
    try:
        selected_headers, feature_headers, target_header, rows, summary = select_features(
            headers=payload.dataset.headers,
            rows=payload.dataset.rows,
            feature_indices=payload.feature_indices,
            target_index=payload.target_index,
            max_rows=payload.max_rows,
        )
    except ValueError as error:
        raise HTTPException(status_code=400, detail=str(error)) from error

    return FeatureSelectionResponse(
        format=payload.dataset.format,
        selected_headers=selected_headers,
        feature_headers=feature_headers,
        target_header=target_header,
        rows=rows,
        summary=summary,
    )
